### PR TITLE
Pretty enums

### PR DESCRIPTION
**Pull request by ciechowoj**
Adds a functionality to remove superfluous prefixes of the enumeration members.

### CODE REVIEW

**Comment by jacob-carlborg** [comment]
I just had a quick look so far:

* For the future, it would be nice with with a small example in the PR description, then I don't need to read the tests to figure out what it's doing. Even better, include the example in the commit message as well
* Is this enabled with a flag or always on?
* How does it decide that this renaming will be used?
* There are a few style issues (like extra newlines)
* There's no entry in the changelog (you should know this by now 😃)

**Comment by ciechowoj** [comment]
Find the answers inline.

On Aug 25, 2017 09:00, "jacob-carlborg" <notifications@github.com> wrote:

I just had a quick look so far:

   - For the future, it would be nice with with a small example in the PR
   description, then I don't need to read the tests to figure out what it's
   doing. Even better, include the example in the commit message as well

Acknowledged.


   - Is this enabled with a flag or always on?

It will be off by default.


   - How does it decide that this renaming will be used?

It uses some heuristic (in general it looks at prefixes and tries to guess
the moat suitable one)...


   - There are a few style issues (like extra newlines)

Im workin on it.


   - There's no entry in the changelog (you should know this by now 😃)

Yeah...



—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/jacob-carlborg/dstep/pull/169#issuecomment-324839704>,
or mute the thread

File: unit_tests/PrettyEnumUnitTests.d [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        auto message = format(formatString, expected, chain([input.front], actual));
+        throw new AssertError(message, file, line);
+    }
+}
+
+unittest
+{
+    assertRefactorEnum(
+        ["PrefixFoo", "PrefixFOO", "PrefixBAR", "PrefixBAZ"],
+        ["PrefixFoo", "FOO", "BAR", "BAZ"]);
```
Comment (jacob-carlborg) [inline unit_tests/PrettyEnumUnitTests.d]: Perhaps I don't understand the heuristics yet, but why is `PrefixFoo` renamed?

**Comment by jacob-carlborg** [comment]
Does it rename the members individually or will it rename only if all members can be renamed?

**Comment by jacob-carlborg** [comment]
This is a pattern used by Apple, mostly with Objective-C:

```objc
#include <Foundation/NSObjCRuntime.h>

typedef NS_ENUM(NSUInteger, NSApplicationDelegateReply) {
    NSApplicationDelegateReplySuccess = 0,
    NSApplicationDelegateReplyCancel = 1,
    NSApplicationDelegateReplyFailure = 2
};
```

This is imported to Swift as:

```swift
enum NSApplicationDelegateReply : NSUInteger {
    case success
    case cancel
    case failure
}
```

Will this PR strip the `NSApplicationDelegateReply` prefix in this example?

Would we like to do the lowercasing as well?

**Comment by ciechowoj** [comment]
The first item in the array is the name of the enum itself that is why it
isn't renamed.

**Comment by ciechowoj** [comment]
It looks at all of the members an the name of the enum.

However the case where there is only one member is handled separately.

**Comment by jacob-carlborg** [comment]
> The first item in the array is the name of the enum itself that is why it
isn't renamed.

Aha, and that is part of the heuristics?

**Comment by ciechowoj** [comment]
Yes.

**Comment by jacob-carlborg** [comment]
Cool.

**Comment by ciechowoj** [comment]
> Would we like to do the lowercasing as well?

Dont know. Maybe as an option?

**Comment by jacob-carlborg** [comment]
Would `rename`, instead of `refactor`, be a better name?

**Comment by ciechowoj** [comment]
I'll consider adding it as the default.

**Comment by ciechowoj** [comment]
I'll think about it.

**Comment by ciechowoj** [comment]
Fixed some corner cases.
Added conversion to camelCase.
Added more unit tests.

File: unit_tests/RenameEnumMembersUnitTests.d [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        auto message = format(formatString, expected, chain([input.front], actual));
+        throw new AssertError(message, file, line);
+    }
+}
+
+unittest
+{
+    assertRenameEnumMembers(
+        ["PrefixFoo", "PrefixFOO", "PrefixBAR", "PrefixBAZ"],
+        ["PrefixFoo", "FOO", "BAR", "BAZ"]);
```
Comment (jacob-carlborg) [inline unit_tests/RenameEnumMembersUnitTests.d]: Wouldn't it be more clear if the enum names (not the members) were passed as separate arguments from the members?

Comment (ciechowoj) [reply unit_tests/RenameEnumMembersUnitTests.d]: Oh, whatever.

**Comment by jacob-carlborg** [comment]
I think this looks good. Could you please squash the commits?

**Comment by ciechowoj** [comment]
Rebased.

**Comment by jacob-carlborg** [comment]
👌.

**Comment by jacob-carlborg** [comment]
Thanks.
